core/remotes/docker/lister.go

```diff
@@ -0,0 +1,144 @@
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package docker
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/containerd/errdefs"
+	"github.com/containerd/log"
+)
+
+var ErrObjectNotRequired = errors.New("object not required")
+
+type TagList struct {
+	Name string   `json:"name"`
+	Tags []string `json:"tags"`
+}
+
+type dockerLister struct {
+	dockerBase *dockerBase
+}
+
+func (r *dockerLister) List(ctx context.Context) ([]string, error) {
+	refSpec := r.dockerBase.refspec
+	base := r.dockerBase
+	var (
+		errs  error
+		paths [][]string
+		caps  = HostCapabilityPull
+	)
+
+	// turns out, we have a valid digest, make an url.
+	paths = append(paths, []string{"tags/list"})
+	caps |= HostCapabilityResolve
+
+	hosts := base.filterHosts(caps)
+	if len(hosts) == 0 {
+		return nil, fmt.Errorf("no list hosts: %w", errdefs.ErrNotFound)
+	}
+
+	ctx, err := ContextWithRepositoryScope(ctx, refSpec, false)
+	if err != nil {
+		return nil, err
+	}
+
+	// right now, this is a single path only. But we can add more variations.
+	for _, u := range paths {
+		for _, host := range hosts {
+			ctxWithLogger := log.WithLogger(ctx, log.G(ctx).WithField("host", host.Host))
+
+			req := base.request(host, http.MethodGet, u...)
+			if err := req.addNamespace(base.refspec.Hostname()); err != nil {
+				return nil, err
+			}
+
+			req.header["Accept"] = []string{"application/json"}
+
+			log.G(ctxWithLogger).Debug("listing")
+
+			data, ok := r.tryHost(ctxWithLogger, req, &errs, host)
+			if !ok {
+				log.G(ctxWithLogger).WithError(err).Debug("trying next host")
+
+				continue
+			}
+
+			tags := &TagList{}
+			if err = json.Unmarshal(data, tags); err != nil {
+				return nil, err
+			}
+
+			return tags.Tags, nil
+		}
+	}
+
+	// If above loop terminates without return, then there was an error.
+	// "firstErr" contains the first non-404 error. That is, "firstErr == nil"
+	// means that either no registries were given or each registry returned 404.
+
+	if errs == nil {
+		errs = fmt.Errorf("%s: %w", base.refspec.Locator, errdefs.ErrNotFound)
+	}
+
+	return nil, errs
+}
+
+// tryHost tries a host and sets the firstErr of we would like to have that set.
+// In case of NotFound we are ignoring the error because we just want to continue and try the next host.
+func (r *dockerLister) tryHost(ctx context.Context, req *request, firstErr *error, host RegistryHost) ([]byte, bool) {
+	resp, err := req.doWithRetries(ctx, nil)
+	if err != nil {
+		if errors.Is(err, ErrInvalidAuthorization) {
+			err = fmt.Errorf("pull access denied, repository does not exist or may require authorization: %w", err)
+		}
```

Comment:
This error message is referring to `pull`. Also slightly wondering if it should be wrapped at all; I _think_ the error  message used here was to provide backward-compatibility with error messages returned by dockerd when pushing / pulling an image, but dockerd does not provide listing tags as a features, so at least from that perspective that wouldn't be needed here.